### PR TITLE
Allow additional test types

### PR DESCRIPTION
**Pull request by zauguin**
Add support for custom test types with individual normalization/processing and potentially independent comparision code.

### CODE REVIEW

**Comment by josephwright** [comment]
Is this a 'work in progress' or do you want comments?

**Comment by zauguin** [comment]
I would appreciate comments, especially about the choice of customization points.  (Sorry for the whitespace changes, I normalized the file in the end  and that picked up some old lines too, but it felt silly to revert that.)

File: l3build-check.lua
Comment (zauguin) [inline l3build-check.lua:564]: This might look a bit odd because there already is `rewrite` which can be customized with a normalize function, for for other formats it is often easier to access a file directly, therefore this additional indirection.

File: l3build-check.lua
Comment (zauguin) [inline l3build-check.lua:678]: One of the bigger customization points. The idea is that this will almost always just be diff which is therefore the default, but it allows more flexibility if necessary. The base_compare wrapper ensures that the interface is relatively simple. In most cases, a compare function can just take three arguments with the output and the to be compared input filenames.

File: l3build-check.lua
Comment (zauguin) [inline l3build-check.lua:714]: It would be a bit nicer to hide `ext` too in `test_type`, but it has to be separate to allow expectation files.

File: l3build-check.lua
Comment (zauguin) [inline l3build-check.lua:825]: This whole loop could be made faster by skipping the directory listing and just checking the filenames directly, but then we woul no longer allow arbitrary patterns in auxfiles so this would be a breaking change.

File: l3build-check.lua
Comment (zauguin) [inline l3build-check.lua:814]: This drops the explicit exclusion of out types, but they really shouldn't be matched as auxfiles anyway.

File: l3build-check.lua
Comment (zauguin) [inline l3build-check.lua:880]: Here we change the logic in the process: Instead of first listing all excluded files, we check for exclusions while scanning. That's a bit cleaner to implement in the generic case and it also saves some directory listings.

File: l3build-check.lua
Comment (zauguin) [inline l3build-check.lua:1042]: I left this error as is for now because I didn't understand what it's supposed to do. It errors if no test file, but an expectation file exists. But why is it talking about overwriting? If I'm not missing something, the saved tlgfile can't overwrite anything because without the testfile we can't even generate a tlg file.

File: l3build-file-functions.lua
Comment (zauguin) [inline l3build-file-functions.lua:67]: I just wanted shorter patterns during debugging so I rewrote the converter and committed it by accident. I will revert this before merging.

File: l3build-variables.lua
Comment (zauguin) [inline l3build-variables.lua:229]: The default configuration. Maybe three interesting parts here: The split between `test_types` and `test_order`. Mostly it's done this way to ensure that behavior is deterministic in some edge cases, but it also allows to a switch to a auto loading schemes for additional formats later (To automatically fill test_types when types are used in test_order).

Names might need improvement.

**Review by zauguin — COMMENTED** [review]
Some inline comments bout the more interesting parts. (Here more interesting means: Not just replacing some variable names or replacing repetition with loops)

File: l3build-check.lua
Comment (josephwright) [inline l3build-check.lua:603]: Do we really want an assertion here? Won't that leave us with an uncontrolled error?

File: l3build-check.lua
Comment (josephwright) [reply l3build-check.lua:814]: Probably this was historical anyway: it looks OK but we need to see what happens when applied to the 2e repo :)

File: l3build-check.lua
Comment (josephwright) [reply l3build-check.lua:1042]: This was written a while ago when things were rather less complicated: I may well have copied the text between two places where the outcomes where different! I agree it doesn't look like it doesn't seem to make sense: change the text to say that?

File: l3build-file-functions.lua
Comment (josephwright) [reply l3build-file-functions.lua:67]: Feel free to change if it works better: I simply took some existing code, simplified a little and used it. I'm not really good with LPEG so it's very likely that there are better ways.

File: l3build-variables.lua
Comment (josephwright) [reply l3build-variables.lua:229]: I'm not clear on what you are up to with metatables here :) Why not `test_types = test_types or  { }` then 
`test_types.log = test_types.log or { ... }` and so on?

File: l3build.dtx
Comment (josephwright) [inline l3build.dtx:1280]: This does not match how everything else works. I'd expect `test_types  =  test_types or { }`, `test_types.mytest = test_types.mytext or { ... }`, the same as we do for 'special' engines and so on.

Comment (josephwright) [reply l3build.dtx:1280]: What happens if only some entries are given? Do we need a prototype to be copied then altered, for example?

**Review by josephwright — CHANGES_REQUESTED** [review]
Looks good overall but there are a few places I'd like a bit of clarification (no code changes required necessarily).

File: l3build-check.lua
Comment (zauguin) [reply l3build-check.lua:814]: 2e does not change auxfiles, therefore ext can never have these values anyway.

File: l3build-check.lua
Comment (zauguin) [reply l3build-check.lua:1042]: I would just drop this branch. It only triggers when a `lve` file exists but no test and then `l3build save` is called. While the `lve` file is odd, the real issue we should be reporting is the missing test.

File: l3build.dtx
Comment (zauguin) [reply l3build.dtx:1280]: I don't really understand what you mean here.

All the entries `test`, `reference` and `generated` are required because otherwise the system doesn't know which files to operate on. Similarly, `rewrite` is effectivly always required because there is no reasonable default (We could copy without normalization but that would never be reproducible.)

The case of copying and altering a prototype sounds like you would just have two parts in `rewrite`: First copy, then rewrite. (Or both steps can be combined as in the default normalization functions)

File: l3build-check.lua
Comment (zauguin) [reply l3build-check.lua:603]: It doesn't do anything because `errlevel` can never be `nil` here. (I had the assert to ensure that I didn't miss a case where it can happen, but I'll take it out again.

File: l3build-check.lua
Comment (zauguin) [reply l3build-check.lua]: @josephwright The new version now moves the error to actually warn if a lve is present while trying to save a test.

File: l3build-variables.lua
Comment (zauguin) [reply l3build-variables.lua]: @josephwright No real reason, I'm just more used to the metatable approach. I changed it for consistency now.

**Comment by josephwright** [comment]
I think modulo a few merge issues we are good to go here. I think also that with one or two more tweaks this would allow @jlaurens to test Lua - we just need to make sure that `\input` is not added in those cases, which we can do by adding some entry to the setup table.

Anyone else got comments (@cereda ?)

**Comment by jlaurens** [comment]
The reference/comparison model does not fit with unit testing without a very very significant overhead. Some work must be done before attempting to use this route, so do not feel in a hurry for me. This types enhancement makes sense on its own.

In my opinion, the main remark is that unit testing has nothing to copy to the build directory.
So there is no reference, no expectation, no normalisation and nothing to rewrite. The only thing that makes a difference seems to be the `compare` method. I have difficulties to imagine a straightforward workflow for this approach.

However, the build directory might be used as a playground to test file functions on temp files and folders.

BTW, some how to recipe may help in following the path, but that is not a requirement of mine.

**Comment by josephwright** [comment]
> The reference/comparison model does not fit with unit testing without a very very significant overhead. Some work must be done before attempting to use this route, so do not feel in a hurry for me. This types enhancement makes sense on its own.

Indeed: I'd mainly got distracted from looking over the PR here. It's a great idea, we just have be sure we've got the interface about right. I plan to check that partly by setting up some Lua testing in a dependent branch, and making sure it looks OK.

> In my opinion, the main remark is that unit testing has nothing to copy to the build directory.

I'm afraid I don't see that. The reason we move everything for testing is that is the only way to have an 'isolated' system in which one can be sure there are not stray accessible files. Copying everything to a working space ensures there are no 'oddities'. At the same time, other than `l3build` itself and a few special cases (`luaotfload` is the obvious one), I'd expect Lua projects in the 'TeX sphere' to be packaged as `.dtx` sources or similar, so there's the need to extract using DocStrip first. That's a lot cleaner done in a 'disposable' space.

> So there is no reference, no expectation, no normalisation and nothing to rewrite. The only thing that makes a difference seems to be the `compare` method. I have difficulties to imagine a straightforward workflow for unit testing with this approach.

Seems simple to me. Write a test file (could be as `.lua`), output the results to a file (say `.log`), save a standardised version as say `.llg`. So more-or-less the same as for LaTeX, just with different extensions. That might be supported by some `regression-test.lua` in the style of `regression-test.tex`, but could be a lot less complex (no real need to worry about normalisation after-the-fact in most cases).

**Comment by jlaurens** [comment]
You are still far ahead of me. You are more talking about adding new regression testing capabilities to l3build whereas I am focused on developing l3build with the help of unit testing.
So your examples do make sense from your purpose but not for mine.

**Comment by josephwright** [comment]
> You are still far ahead of me. You are more talking about adding new regression testing capabilities to l3build whereas I am focused on developing l3build with the help of unit testing.
> So your examples do make sense from your purpose but not for mine.

Huh? You can happily do something akin to
```lua
require("some-module")

if some.thing then
  TESTOUT("All good")
else
   TESTOUT("Fail")
end
```
using some `TESTOUT` that writes to e.g. a `.log` file.

**Comment by jlaurens** [comment]
simple on the paper. On everyday use, one gets rapidly into trouble.
Not to mention the testing aspect that must be practical (no intermediate text file, simple code), just trying to
```
require `l3build`
```
is actually difficult. One of these is required upwind
- redefine main
- define global options and bypass option parsing
- pass options to the testing script
- use os.execute with options instead of require
That sounds a bit complicated for a testing technique. Anyway, it does not give fine grain control.

I easily imagine that testing "arguments", "variables" or "file functions" module alone is rather easy.
To test "help", both "arguments" and "stdmain" are required but I guess the tests would be ok without the rest.
For the others, I am less optimistic because the tests must be made in a global context.
This global context comes from arguments parsing and variables management.
Both can be simulated in theory, but does It worth the cost?

**Comment by josephwright** [comment]
@zauguin  Could you put in the LPEG stuff as a separate PR? That would make things nice and clean in many ways, but is separate from the main part here.

**Comment by zauguin** [comment]
@josephwright Done. I also used the this to rebase on current master.

**Comment by josephwright** [comment]
@zauguin  Er, don't see a separate PR!

**Comment by zauguin** [comment]
@josephwright Any other comments or can we merge this after resolving the conflicts?

**Comment by josephwright** [comment]
I think we are good to go

**Pull request merged**
